core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala

```
@@ -0,0 +1,116 @@
+import org.apache.kafka.clients.producer.{ProducerConfig, ProducerRecord}
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.utils.Utils
+import org.apache.kafka.common.errors.{KafkaStorageException, NotLeaderForPartitionException}
+import org.junit.{Before, Test}
+import org.junit.Assert.assertTrue
+
+/**
+  * Test whether clients can producer and consume when there is log directory failure
+  */
+class LogDirFailureTest extends IntegrationTestHarness {
+  val producerCount: Int = 1
+  val consumerCount: Int = 1
+  val serverCount: Int = 2
+  private val topic = "topic"
+
+  this.logDirCount = 2
+  this.producerConfig.setProperty(ProducerConfig.RETRIES_CONFIG, "0")
+  this.producerConfig.setProperty(ProducerConfig.RETRY_BACKOFF_MS_CONFIG, "100")
+  this.serverConfig.setProperty(KafkaConfig.ReplicaHighWatermarkCheckpointIntervalMsProp, "100")
+
+  @Before
+  override def setUp() {
+    super.setUp()
+    TestUtils.createTopic(zkUtils, topic, 1, 2, servers = servers)
+  }
+
+  @Test
+  def testProduceAfterLogDirFailure() {
+
+    val consumer = consumers.head
+    subscribeAndWaitForAssignment(topic, consumer)
+    val producer = producers.head
+    val partition = new TopicPartition(topic, 0)
+    val record = new ProducerRecord(topic, 0, s"key".getBytes, s"value".getBytes)
+
+    val leaderServerId = producer.partitionsFor(topic).get(0).leader().id()
+    val leaderServer = servers.find(_.config.brokerId == leaderServerId).get
+
+    // The first send() should succeed
+    producer.send(record).get()
+    TestUtils.waitUntilTrue(() => {
+      consumer.poll(0).count() == 1
+    }, "Expected the first message", 3000L)
+
+    // Make log directory of the partition on the leader broker inaccessible by replacing it with a file
+    val replica = leaderServer.replicaManager.getReplica(partition)
+    val logDir = replica.get.log.get.dir.getParentFile
+    CoreUtils.swallow(Utils.delete(logDir))
+    logDir.createNewFile()
```

Could we just set the file to be unreadable and unwritable?

---

I just tested it by setting the file to be unwritable. The testProduceAfterLogDirFailure() with pass. However, `KafkaServerTestHarness.tearDown()` will fail with `java.nio.file.AccessDeniedException` because it is not able to delete the log directory. It is possible to swallow the exception so that the test will pass. But that the temporary log directory will not be removed after the test and will accumulate over time on the test server.

According to the information provided below, it seems that the log directory can only be removed with sudo access if it is set to be unwritable. Thus it seems simpler to just use the current approach by replacing the log directory with a file.

```
[dolin@dolin-ld2 kafka]$ rm -rf /tmp/kafka-1650964778799526893/log-start-offset-checkpoint
rm: cannot remove `/tmp/kafka-1650964778799526893/log-start-offset-checkpoint': Permission denied
[dolin@dolin-ld2 kafka]$ ls -ld /tmp/kafka-1650964778799526893/log-start-offset-checkpoint
-rw-r--r-- 1 dolin eng 0 Jul 20 15:08 /tmp/kafka-1650964778799526893/log-start-offset-checkpoint
```